### ISSUE TITLE
Dimmer functionality for Busch-Jaeger 673x

### ISSUE THREAD

**Issue by afoeder** [Koenkk/zigbee-herdsman-converters#2427]
Hi,

apparently the dimmer functionality for the 6715 U (Dimmer) is not yet supported according to the [comment there](https://github.com/Koenkk/zigbee-herdsman-converters/blob/e534456e6ff711acf4d8fbd7e365d307c623e54a/devices.js#L10642).
I own the respective device and would like to have the support for it; can I contribute something?
I must however admit that you'd kinda need to hold my hand because I'd have no idea what I'm doing. But if I could sniff stuff etc. I can gladly help.

I used to use the device directly paired with a Philips Hue bridge as well as deCONZ and in both cases dimming was possible (maybe this fact helps reverse-engineering).

/cc @docbobo 

Thank you guys!

**Comment by afoeder** [comment]
Hi @docbobo do you think I could contribute something here? However I'd need some onboarding help I guess.

**Comment by docbobo** [comment]
I don't see why you shouldn't - this is open source, isn't it? ;) However, I don't exactly know what's needed here. Anyway, let's dump some of my memories here:

- The devices are limited in memory. I suspect that implementing dimming functionality might run into its limits because you'll probably have to bind to additional clusters. Just a guess though.
- I am not sure how easy it will be to differentiate between a dimmer and a regular switch. If you don't want to break existing installations, that need to be addressed.
- You might want to look at different dimmer implementations for inspiration. The last time I checked, they were mostly starting some form of timer on a button down event, stopping it on button up. However, I believe the codebase has changed ever since, so maybe this is much easier than in the past.

**Comment by afoeder** [comment]
So, please help me out to understand. To keep it simple for me, let's assume the relay item (6711 U) and a control panel.
What happens zigbee-wise when I add it to a Hue bridge network? What I can see is: there is a switch available, when pressing the top row the relay will be switched on/off respectively; and when I switch it on/off in the App, the relay is acted as well.
What happens here? is it…

```plantuml
@startuml
actor User

User -> Panel: Press Key

group Is it…

    Panel --> Relay: Switch On!
    Relay -> Relay: *click*
    Panel --> Bridge: Send state "On"

else or is it…
    
    Panel --> Bridge: Send act? "On"
    Bridge --> Panel: Switch "On"!
    Panel --> Relay: Switch On!
    Relay -> Relay: *click*

end
@enduml
```

![image](https://user-images.githubusercontent.com/1125168/117001212-5a26a300-ace2-11eb-9565-ec11fec49b79.png)

tl;dr: does the panel switch the relay "directly" without any zigbee involved? or does the panel just send the cmd to the bridge (i.e. the coordinator) and the bridge sends a cmd to the relay? The latter would be congruent with a user pushing a button in the app.

I guess the same would apply for the dimmer then, but that's another story I'll form when I've understood the above. Thank you!

**Comment by docbobo** [comment]
It's my understanding that the panel directly controls either relays or dimmer directly, without the zigbee network being involved. That's also the reason, why the coordinator usually doesn't reflect the correct state after you press the switch. 

But you wanted to start simple and I realized that I had made some assumptions here. What exactly is it you'd like to achieve here? I would have assumed that it should be possible to control the dimmer directly via the hardware button. When I wrote my comment, I assumed that you'd want to use other rows of the panel to have a similar effect on other devices, but it seems I was wrong about that.

**Comment by afoeder** [comment]
>panel directly controls either relays or dimmer directly

right, that's also my assumption.

Anyways to the dimmer, and back to z2m: currently when I use the 6715 (i.e. the dimmer), Z2m doesn't expose the dimming capability! At least not fully…
In Z2M there's only the (binary) state:
![image](https://user-images.githubusercontent.com/1125168/117016584-24d68100-acf3-11eb-8e9d-4a4e7a37c4dc.png)

However in the State tab, there is a `brightness_relay` item which indeed correlates with the current brightness:
![image](https://user-images.githubusercontent.com/1125168/117016716-43d51300-acf3-11eb-9a84-71b688102ea6.png)

In HomeAssistant it's considered a switch:
![image](https://user-images.githubusercontent.com/1125168/117016438-007aa480-acf3-11eb-8edb-2faec1b8524c.png)

Unfortunately I don't have another dimming device to compare to; but I remember when I've paired the Busch-Jaeger with vanilla Hue bridge it appeared as dimmable switch there and was dimmable through the App. To gain this behavior with z2m as well was my intention. TBH I am meanwhile not sure if I took the wrong road at some point in my thinking…

**Comment by afoeder** [comment]
OK, having looked again at the code I have particular questions about Lines [10653-10674](https://github.com/Koenkk/zigbee-herdsman-converters/blob/2dad9b9921238b41d26e6a8ca6e1dba734bd3b29/devices.js#L10653-L10674):

```javascript
endpoint: (device) => {
    // what do these hex numbers refer to? something on the zigbee device?
    // where are these numbers known from? and what does `relay` mean?
    return {'row_1': 0x0a, 'row_2': 0x0b, 'row_3': 0x0c, 'row_4': 0x0d, 'relay': 0x12};
},
exposes: [e.switch(), e.action(['row_1_on', 'row_1_off', 'row_1_up', 'row_1_down', 'row_1_stop',
    'row_2_on', 'row_2_off', 'row_2_up', 'row_2_down', 'row_2_stop',
    'row_3_on', 'row_3_off', 'row_3_up', 'row_3_down', 'row_3_stop',
    'row_4_on', 'row_4_off', 'row_4_up', 'row_4_down', 'row_4_stop'])],
meta: {configureKey: 3, multiEndpoint: true},
configure: async (device, coordinatorEndpoint, logger) => {
    let firstEndpoint = 0x0a;

    // what does the following do? I would assume it's some checking about
    // the first row's capabilities (i.e. is the first row dedicated for switching/dimming
    // like 6711 (relay) and 6715 (dimmer) or not, like 6710; but that interferes with my
    // understanding of the next block…
    const switchEndpoint10 = device.getEndpoint(10);
    if (switchEndpoint10 != null && switchEndpoint10.supportsOutputCluster('genOnOff')) {
        // https://github.com/Koenkk/zigbee2mqtt/issues/3027#issuecomment-606169628
        await reporting.bind(switchEndpoint10, coordinatorEndpoint, ['genOnOff']);
    }

    // in contrary to the former block, what is 0x12? in the `endpoint` key above, it's something
    // with "relay". But what is it compared to Endpoint `10` like in the former block?
    // the rest of the code seems self-explanatory: 
    const switchEndpoint12 = device.getEndpoint(0x12);
    if (switchEndpoint12 != null) { // if the device… has a relay? what does "having 0x12" tell us?
                                    // anyways it means if the top row is dedicated,
        firstEndpoint++;            // then skip the first row in the block ahead
        await reporting.bind(       // and bind this dedicated row with switch and level ctrl capabilities.
                switchEndpoint12,
                coordinatorEndpoint, 
                ['genOnOff', 'genLevelCtrl']);
    }
```

maybe you could shed some light on it. Thanks again!

**Comment by docbobo** [comment]
I can help with some of this, not all 10664-10668 were introduced by someone else and the linked comment doesn't seem to be particularly helpful.

When configure is called, the BJ device will provide a list of endpoints and capabilities. If you are running a relay (and maybe a dimmer), it will announce EP 18 in addition to 10, 11, 12, 13. If EP 18 is announced, EP 10 has no value because the first row of the switch is directly controlling the relay.

Hope this makes some sense.
